File: core/ffmpeg.py
"""Wrapper tipado para ``ffmpeg`` e ``ffprobe``.

Toda operação de vídeo e áudio passa por aqui. O restante do código nunca
monta linha de comando nem interpreta saída bruta.
"""

from __future__ import annotations

import json
import shutil
import subprocess
from dataclasses import dataclass
from fractions import Fraction
from typing import TYPE_CHECKING, Final, TypedDict

from core.errors import ToolError

if TYPE_CHECKING:
    from collections.abc import Sequence
    from pathlib import Path

_STDERR_TAIL: Final = 1500


class ProbeResult(TypedDict):
    """Metadados essenciais de um arquivo de mídia."""

    duration: float
    width: int | None
    height: int | None
    fps: float | None
    video_codec: str | None
    audio_codec: str | None
    has_video: bool
    has_audio: bool
    size_bytes: int
    format_name: str


@dataclass(frozen=True, slots=True)
class FFmpeg:
    """Executa binários do FFmpeg com timeout e erros amigáveis."""

    ffmpeg_bin: str = "ffmpeg"
    ffprobe_bin: str = "ffprobe"
    timeout_seconds: float = 600.0

    def is_available(self) -> bool:
        """Indica se os dois binários estão no PATH."""
        return (
            shutil.which(self.ffmpeg_bin) is not None and shutil.which(self.ffprobe_bin) is not None
        )

    def require(self) -> None:
        """Falha com mensagem clara se o FFmpeg não estiver instalado.

        Raises:
            ToolError: Se ffmpeg ou ffprobe não forem encontrados.
        """
        if not self.is_available():
            raise ToolError(
                "ffmpeg/ffprobe não encontrados no PATH.",
                code="unavailable",
                hint="Instale o FFmpeg ou ajuste FFMPEG_BIN e FFPROBE_BIN.",
            )

    def run(self, args: Sequence[str]) -> str:
        """Roda ``ffmpeg`` com os argumentos dados e devolve o stderr.

        O stderr é onde o FFmpeg escreve progresso e informações de filtros.

        Raises:
            ToolError: Se o processo falhar ou exceder o timeout.
        """
        self.require()
        cmd = [self.ffmpeg_bin, "-hide_banner", "-nostdin", "-y", *args]
        try:
            proc = subprocess.run(
                cmd,
                capture_output=True,
                text=True,
                timeout=self.timeout_seconds,
                check=False,
            )
        except subprocess.TimeoutExpired as exc:
            raise ToolError(
                f"ffmpeg excedeu o timeout de {self.timeout_seconds:.0f}s.",
                code="timeout",
                hint="Use background=True para operações longas ou aumente FFMPEG_TIMEOUT.",
            ) from exc
        if proc.returncode != 0:
            raise ToolError(
                f"ffmpeg falhou (código {proc.returncode}): {proc.stderr[-_STDERR_TAIL:].strip()}",
                code="ffmpeg_failed",
                hint="Verifique se o arquivo de entrada é válido com probe_video.",
            )
        return proc.stderr

    def probe(self, path: Path) -> ProbeResult:
        """Lê metadados de um arquivo de mídia via ``ffprobe``.

        Raises:
            ToolError: Se o ffprobe falhar ou a saída não puder ser interpretada.
        """
        self.require()
        cmd = [
            self.ffprobe_bin,
            "-v",
            "error",
            "-print_format",
            "json",
            "-show_format",
            "-show_streams",
            str(path),
        ]
        try:
            proc = subprocess.run(
                cmd,
                capture_output=True,
                text=True,
                timeout=self.timeout_seconds,
                check=False,
            )
        except subprocess.TimeoutExpired as exc:
            raise ToolError("ffprobe excedeu o timeout.", code="timeout") from exc
        if proc.returncode != 0:
            raise ToolError(
                f"ffprobe falhou: {proc.stderr[-_STDERR_TAIL:].strip()}",
                code="ffmpeg_failed",
                hint="O arquivo pode estar corrompido ou não ser um vídeo/áudio.",
            )
        return _parse_probe(proc.stdout, path)


def _parse_probe(raw: str, path: Path) -> ProbeResult:
    try:
        data: dict[str, object] = json.loads(raw)
    except json.JSONDecodeError as exc:
        raise ToolError("Saída do ffprobe não é JSON válido.", code="ffmpeg_failed") from exc

    fmt = data.get("format")
    streams = data.get("streams")
    if not isinstance(fmt, dict) or not isinstance(streams, list):
        raise ToolError("Saída do ffprobe sem 'format' ou 'streams'.", code="ffmpeg_failed")

    video = next(
        (s for s in streams if isinstance(s, dict) and s.get("codec_type") == "video"), None
    )
    audio = next(
        (s for s in streams if isinstance(s, dict) and s.get("codec_type") == "audio"), None
    )

    return ProbeResult(
        duration=_as_float(fmt.get("duration")) or 0.0,
        width=_as_int(video.get("width")) if video else None,
        height=_as_int(video.get("height")) if video else None,
        fps=_as_fps(video.get("avg_frame_rate")) if video else None,
        video_codec=_as_str(video.get("codec_name")) if video else None,
        audio_codec=_as_str(audio.get("codec_name")) if audio else None,
        has_video=video is not None,
        has_audio=audio is not None,
        size_bytes=_as_int(fmt.get("size")) or path.stat().st_size,
        format_name=_as_str(fmt.get("format_name")) or "",
    )


def _as_float(value: object) -> float | None:
    try:
        return float(value) if isinstance(value, (str, int, float)) else None
    except ValueError:
        return None


def _as_int(value: object) -> int | None:
    try:
        return int(value) if isinstance(value, (str, int)) else None
    except ValueError:
        return None


def _as_str(value: object) -> str | None:
    return value if isinstance(value, str) else None


def _as_fps(value: object) -> float | None:
    if not isinstance(value, str) or value in {"", "0/0"}:
        return None
    try:
        return round(float(Fraction(value)), 3)
    except ValueError, ZeroDivisionError:
        return None

File: core/ffmpeg.rs
//! Wrapper tipado para `ffmpeg` e `ffprobe`.
//!
//! Toda operação de vídeo e áudio passa por aqui. O restante do código nunca
//! monta linha de comando nem interpreta saída bruta.

use std::ffi::OsStr;
use std::path::{Path, PathBuf};
use std::time::Duration;

use serde::{Deserialize, Serialize};
use serde_json::Value;

use crate::core::binaries::{Binaries, ExternalTool};
use crate::core::errors::{ErrorCode, ToolError, ToolResult};
use crate::core::process::{self, ProcessError};

const STDERR_TAIL: usize = 1500;

/// Monta a lista de argumentos do ffmpeg a partir de itens heterogêneos
/// (`&str`, `String`, `&Path`, `PathBuf`...), como a lista do Python.
#[macro_export]
macro_rules! ffargs {
    ($($item:expr),* $(,)?) => {
        vec![$(::std::ffi::OsString::from(::std::convert::AsRef::<::std::ffi::OsStr>::as_ref(&$item))),*]
    };
}

/// Metadados essenciais de um arquivo de mídia.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct ProbeResult {
    pub duration: f64,
    pub width: Option<i64>,
    pub height: Option<i64>,
    pub fps: Option<f64>,
    pub video_codec: Option<String>,
    pub audio_codec: Option<String>,
    pub has_video: bool,
    pub has_audio: bool,
    pub size_bytes: u64,
    pub format_name: String,
}

/// Executa binários do FFmpeg com timeout e erros amigáveis.
#[derive(Debug, Clone)]
pub struct FFmpeg {
    pub ffmpeg_bin: String,
    pub ffprobe_bin: String,
    pub timeout_seconds: f64,
    pub binaries: Binaries,
}

impl Default for FFmpeg {
    fn default() -> Self {
        Self::new("ffmpeg", "ffprobe", 600.0, Binaries::default())
    }
}

impl FFmpeg {
    /// Cria o wrapper com os nomes (ou caminhos) dos binários e o timeout por chamada.
    pub fn new(
        ffmpeg_bin: impl Into<String>,
        ffprobe_bin: impl Into<String>,
        timeout_seconds: f64,
        binaries: Binaries,
    ) -> Self {
        Self {
            ffmpeg_bin: ffmpeg_bin.into(),
            ffprobe_bin: ffprobe_bin.into(),
            timeout_seconds,
            binaries,
        }
    }

    /// Indica se os dois binários já estão disponíveis, sem baixar nada.
    pub fn is_available(&self) -> bool {
        self.binaries
            .locate(ExternalTool::Ffmpeg, &self.ffmpeg_bin)
            .is_some()
            && self
                .binaries
                .locate(ExternalTool::Ffprobe, &self.ffprobe_bin)
                .is_some()
    }

    /// Garante os dois binários, baixando se a configuração permitir.
    ///
    /// # Errors
    ///
    /// [`ErrorCode::Unavailable`] se ffmpeg ou ffprobe não forem encontrados.
    pub fn require(&self) -> ToolResult<(PathBuf, PathBuf)> {
        let ffmpeg = self.binaries.ensure(ExternalTool::Ffmpeg, &self.ffmpeg_bin);
        let ffprobe = self
            .binaries
            .ensure(ExternalTool::Ffprobe, &self.ffprobe_bin);
        match (ffmpeg, ffprobe) {
            (Ok(ffmpeg), Ok(ffprobe)) => Ok((ffmpeg, ffprobe)),
            (Err(error), _) | (_, Err(error)) if error.code == ErrorCode::DownloadFailed => {
                Err(error)
            }
            _ => Err(ToolError::with_hint(
                "ffmpeg/ffprobe não encontrados no PATH.",
                ErrorCode::Unavailable,
                "Instale o FFmpeg, ajuste FFMPEG_BIN e FFPROBE_BIN, ou deixe \
                 MCP_AUTO_DOWNLOAD=true para o servidor baixar sozinho.",
            )),
        }
    }

    fn timeout(&self) -> Duration {
        Duration::from_secs_f64(self.timeout_seconds.max(0.001))
    }

    /// Roda `ffmpeg` com os argumentos dados e devolve o stderr.
    ///
    /// O stderr é onde o FFmpeg escreve progresso e informações de filtros.
    ///
    /// # Errors
    ///
    /// [`ErrorCode::FfmpegFailed`] se o processo falhar, [`ErrorCode::Timeout`]
    /// se exceder o timeout.
    pub fn run<I, S>(&self, args: I) -> ToolResult<String>
    where
        I: IntoIterator<Item = S>,
        S: AsRef<OsStr>,
    {
        let (ffmpeg, _) = self.require()?;
        let mut full: Vec<std::ffi::OsString> = ffargs!["-hide_banner", "-nostdin", "-y"];
        full.extend(args.into_iter().map(|a| a.as_ref().to_os_string()));
        let output = match process::run(&ffmpeg, &full, self.timeout()) {
            Ok(output) => output,
            Err(ProcessError::Timeout(_)) => {
                return Err(ToolError::with_hint(
                    format!("ffmpeg excedeu o timeout de {:.0}s.", self.timeout_seconds),
                    ErrorCode::Timeout,
                    "Use background=true para operações longas ou aumente FFMPEG_TIMEOUT.",
                ))
            }
            Err(ProcessError::Spawn(error)) => {
                return Err(ToolError::with_hint(
                    format!("ffmpeg não pôde ser executado: {error}"),
                    ErrorCode::Unavailable,
                    "Confira FFMPEG_BIN ou apague a pasta de cache para baixar de novo.",
                ))
            }
        };
        if !output.success() {
            return Err(ToolError::with_hint(
                format!(
                    "ffmpeg falhou (código {}): {}",
                    output
                        .status_code
                        .map_or("?".to_string(), |c| c.to_string()),
                    tail(&output.stderr, STDERR_TAIL)
                ),
                ErrorCode::FfmpegFailed,
                "Verifique se o arquivo de entrada é válido com probe_video.",
            ));
        }
        Ok(output.stderr)
    }

    /// Lê metadados de um arquivo de mídia via `ffprobe`.
    ///
    /// # Errors
    ///
    /// [`ErrorCode::FfmpegFailed`] se o ffprobe falhar ou a saída não puder ser
    /// interpretada; [`ErrorCode::Timeout`] se exceder o timeout.
    pub fn probe(&self, path: &Path) -> ToolResult<ProbeResult> {
        let (_, ffprobe) = self.require()?;
        let args = ffargs![
            "-v",
            "error",
            "-print_format",
            "json",
            "-show_format",
            "-show_streams",
            path
        ];
        let output = match process::run(&ffprobe, &args, self.timeout()) {
            Ok(output) => output,
            Err(ProcessError::Timeout(_)) => {
                return Err(ToolError::new(
                    "ffprobe excedeu o timeout.",
                    ErrorCode::Timeout,
                ))
            }
            Err(ProcessError::Spawn(error)) => {
                return Err(ToolError::with_hint(
                    format!("ffprobe não pôde ser executado: {error}"),
                    ErrorCode::Unavailable,
                    "Confira FFPROBE_BIN ou apague a pasta de cache para baixar de novo.",
                ))
            }
        };
        if !output.success() {
            return Err(ToolError::with_hint(
                format!("ffprobe falhou: {}", tail(&output.stderr, STDERR_TAIL)),
                ErrorCode::FfmpegFailed,
                "O arquivo pode estar corrompido ou não ser um vídeo/áudio.",
            ));
        }
        parse_probe(&output.stdout, path)
    }
}

/// Últimos `limit` caracteres de `text`, sem espaços nas pontas.
fn tail(text: &str, limit: usize) -> String {
    let count = text.chars().count();
    text.chars()
        .skip(count.saturating_sub(limit))
        .collect::<String>()
        .trim()
        .to_string()
}

/// Interpreta o JSON do `ffprobe`. Público para os testes, como `_parse_probe`.
///
/// # Errors
///
/// [`ErrorCode::FfmpegFailed`] se o JSON for inválido ou sem `format`/`streams`.
pub fn parse_probe(raw: &str, path: &Path) -> ToolResult<ProbeResult> {
    let data: Value = serde_json::from_str(raw).map_err(|_| {
        ToolError::new(
            "Saída do ffprobe não é JSON válido.",
            ErrorCode::FfmpegFailed,
        )
    })?;
    let (Some(fmt), Some(streams)) = (
        data.get("format").and_then(Value::as_object),
        data.get("streams").and_then(Value::as_array),
    ) else {
        return Err(ToolError::new(
            "Saída do ffprobe sem 'format' ou 'streams'.",
            ErrorCode::FfmpegFailed,
        ));
    };
    let stream_of = |kind: &str| {
        streams
            .iter()
            .filter_map(Value::as_object)
            .find(|s| s.get("codec_type").and_then(Value::as_str) == Some(kind))
    };
    let video = stream_of("video");
    let audio = stream_of("audio");
    let size_bytes = as_int(fmt.get("size"))
        .and_then(|v| u64::try_from(v).ok())
        .or_else(|| std::fs::metadata(path).ok().map(|m| m.len()))
        .unwrap_or(0);
    Ok(ProbeResult {
        duration: as_float(fmt.get("duration")).unwrap_or(0.0),
        width: video.and_then(|v| as_int(v.get("width"))),
        height: video.and_then(|v| as_int(v.get("height"))),
        fps: video.and_then(|v| as_fps(v.get("avg_frame_rate"))),
        video_codec: video.and_then(|v| as_str(v.get("codec_name"))),
        audio_codec: audio.and_then(|a| as_str(a.get("codec_name"))),
        has_video: video.is_some(),
        has_audio: audio.is_some(),
        size_bytes,
        format_name: as_str(fmt.get("format_name")).unwrap_or_default(),
    })
}

fn as_float(value: Option<&Value>) -> Option<f64> {
    match value? {
        Value::Number(n) => n.as_f64(),
        Value::String(s) => s.trim().parse().ok(),
        _ => None,
    }
}

fn as_int(value: Option<&Value>) -> Option<i64> {
    match value? {
        Value::Number(n) => n.as_i64(),
        Value::String(s) => s.trim().parse().ok(),
        _ => None,
    }
}

fn as_str(value: Option<&Value>) -> Option<String> {
    value.and_then(Value::as_str).map(str::to_string)
}

fn as_fps(value: Option<&Value>) -> Option<f64> {
    let raw = value?.as_str()?;
    if raw.is_empty() || raw == "0/0" {
        return None;
    }
    let fps = match raw.split_once('/') {
        Some((num, den)) => {
            let num: f64 = num.trim().parse().ok()?;
            let den: f64 = den.trim().parse().ok()?;
            if den == 0.0 {
                return None;
            }
            num / den
        }
        None => raw.trim().parse().ok()?,
    };
    Some((fps * 1000.0).round() / 1000.0)
}
